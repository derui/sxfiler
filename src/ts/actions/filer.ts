import { AppAction, ActionTypes } from "./type";
import { Filer } from "@/domains/filer";
import { Side } from "@/states/file-list";

type SelectAction = AppAction<ActionTypes.FILER_SELECT, { side: Side; itemId: string }>;
type ReloadAction = AppAction<
  ActionTypes.FILER_RELOAD,
  {
    payload: {
      filers: [Filer, Filer];
    };
  }
>;

type UpdateFilerAction = AppAction<
  ActionTypes.FILER_UPDATE,
  {
    payload: {
      filer: Filer;
    };
  }
>;

// The action to load or get a filer from server
type LoadFilerAction = AppAction<
  ActionTypes.FILER_LOAD,
  {
    payload: {
      filer: Filer;
    };
  }
>;

type ChangeSideAction = AppAction<ActionTypes.FILER_CHANGE_SIDE>;

export type Actions = ReloadAction | UpdateFilerAction | ChangeSideAction | LoadFilerAction | SelectAction;

export const reload = function reload(args: { filers: [Filer, Filer] }): ReloadAction {
  return { type: ActionTypes.FILER_RELOAD, payload: { ...args } };
};

/**
 * Update filer
 */
export const update = function update(args: { filer: Filer }): UpdateFilerAction {
  return { type: ActionTypes.FILER_UPDATE, payload: { ...args } };
};

/**
 * Load or get filer on the side.
 */
export const load = function load(args: { filer: Filer }): LoadFilerAction {
  return { type: ActionTypes.FILER_LOAD, payload: { ...args } };
};

/**
   change current side
*/
export const changeSide = function changeSide(): ChangeSideAction {
  return { type: ActionTypes.FILER_CHANGE_SIDE };
};

export const select = function select(side: Side, itemId: string): SelectAction {
  return { type: ActionTypes.FILER_SELECT, side, itemId };
};
