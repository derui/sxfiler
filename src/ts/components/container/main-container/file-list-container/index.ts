export * from "./file-list-container";
