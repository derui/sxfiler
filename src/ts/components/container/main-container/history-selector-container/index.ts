export * from "./history-selector-container";
