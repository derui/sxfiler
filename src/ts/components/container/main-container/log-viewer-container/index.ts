export * from "./log-viewer-container";
