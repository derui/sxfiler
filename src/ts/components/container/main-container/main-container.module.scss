@import "../../ui/commom-style/_colors.scss";
@import "../../ui/commom-style/_variables.scss";
@import "../../ui/commom-style/reset.module.css";

.root {
  max-width: 100%;
  min-width: 0;
  max-height: 100%;
  height: 100%;
  margin: 0;
  padding: 0;

  overflow: hidden;
  background-color: $sol_base03;

  display: grid;
  grid-template-rows: 90% 0 10%;
}
