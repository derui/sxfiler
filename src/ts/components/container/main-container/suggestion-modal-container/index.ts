export * from "./suggestion-modal-container";
