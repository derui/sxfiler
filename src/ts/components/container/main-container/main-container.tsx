import * as React from "react";

import * as Element from "../../ui/element/element";
import * as FileListContainer from "./file-list-container/file-list-container";
import * as NotificationContainer from "./notification-container/notification-container";
import { Component as RootRef } from "../../ui/root-ref/root-ref";

import { ContextLike } from "../../../context";
import LocatorContext from "../../../locator";
import { StoreState } from "../../../types/store-state";
import { State as KeymapState, findBinding } from "../../../types/store-state/keymap";
import * as kbd from "../../../libs/kbd";

const styles = require("./main-container.module.scss");

export interface Props {
  state: StoreState;
}

/**
 * handle keyboard event that all keydown event on application
 * @param props properties of component
 */
function handleKeyDown(context: ContextLike | undefined, keymap: KeymapState): (ev: React.KeyboardEvent<any>) => void {
  return ev => {
    if (!context) {
      return;
    }

    switch (ev.type) {
      case "keydown": {
        const key = kbd.make(ev.key, { meta: ev.metaKey, ctrl: ev.ctrlKey });
        const binding = findBinding(keymap, kbd.toKeySeq(key));

        if (!binding) {
          break;
        }

        ev.preventDefault();
        ev.stopPropagation();
        // TODO: implement to invoke action
        /* onAction(binding.action); */
        break;
      }
      default:
        break;
    }

    return;
  };
}

export class Component extends React.Component<Props> {
  private layoutRef: React.RefObject<HTMLElement> = React.createRef();

  public componentDidMount() {
    if (this.layoutRef.current) {
      this.layoutRef.current.focus();
    }
  }

  public componentDidUpdate() {
    if (this.layoutRef.current) {
      this.layoutRef.current.focus();
    }
  }

  public render() {
    const { fileList, notification, keymap } = this.props.state;
    return (
      <LocatorContext.Consumer>
        {({ context }) => (
          <RootRef rootRef={this.layoutRef}>
            <Element.Component className={styles.root} tabIndex={0} onKeyDown={handleKeyDown(context, keymap)}>
              <FileListContainer.Component key="filer" state={fileList} />
              <NotificationContainer.Component key="notification" state={notification} />
            </Element.Component>
          </RootRef>
        )}
      </LocatorContext.Consumer>
    );
  }
}
