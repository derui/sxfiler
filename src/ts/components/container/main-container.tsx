import { h } from "preact";
import classnames from "classnames";

import * as FilerContainer from "./filer-container";
import * as LogViewerContainer from "./log-viewer-container";
import * as CompleterContainer from "./completer-container";
import * as ConfigurationEditor from "./configuration-editor-container";

import { State } from "@/modules";
import * as DecisionModalContainer from "./decision-modal-container";
import { useTheme } from "@/theme";
import { useContext } from "preact/hooks";
import { LocatorContext } from "@/locator";

export type Props = {
  state: State;
};

export const Component: preact.FunctionComponent<Props> = ({ state }) => {
  const { filer } = state;
  const theme = useTheme();
  const locator = useContext(LocatorContext);

  return (
    <div
      class={classnames({
        "main-container__root ": true,
        [theme || ""]: !!theme,
      })}
      data-testid="mainContainer"
    >
      <FilerContainer.Component state={filer} />
      <DecisionModalContainer.Component state={state} />
      <LogViewerContainer.Component state={state} />
      <CompleterContainer.Component state={state} />
      <ConfigurationEditor.Component state={state} locator={locator} />
    </div>
  );
};
