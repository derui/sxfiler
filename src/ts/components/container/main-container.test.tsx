import { h } from "preact";
import { render } from "preact-render-to-string";

import { State } from "@/modules";
import { Component } from "./main-container";
import { LocatorContext } from "@/locator";
import { emptyState } from "@/modules/filer";

describe("Container", () => {
  describe("Main Container", () => {
    it("render correctly when context not initialized and empty state", () => {
      const locator = {};
      const state = {
        filer: emptyState,
      } as State;

      const wrapper = render(
        <LocatorContext.Provider value={locator}>
          <Component state={state} />
        </LocatorContext.Provider>
      );

      expect(wrapper).toMatchSnapshot();
    });
  });
});
