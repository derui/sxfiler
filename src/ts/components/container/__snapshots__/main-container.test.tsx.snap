// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container Main Container render correctly when context not initialized and empty state 1`] = `"<div class=\\"main-container__root \\" data-testid=\\"mainContainer\\"><div class=\\"log-viewer-container__root\\" data-testid=\\"logViewerContainer\\"><div class=\\"log-viewer__root\\" data-testid=\\"logViewer-root\\"></div></div></div>"`;
